### PR TITLE
Enforce required `PluginParameters`

### PR DESCRIPTION
**Pull request by capricornunicorn123**
### Prerequisites

- [ ] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

Developers have the ability to set a required attribute on a `PluginParameter`. Currently when a `PluginParameter` is displayed in a `PluginParametersSwingDialog` this required attribute simply sets a tooltop and adds "*" to the parameter name. 

This PR deactivates the acceptance button if required parameters are not satisfied. 

### Alternate Designs

The current design of executing a plugin, with missing fields caught during plugin execution, was undesirable as users have to re-execute the plugin from the start rather than getting parameters correct the first time. 

A similar approach could be considered in the Data Access View. Currently plugins can be executed with missing fields and simply execute quietly without performing any graph manipulation. Users have to check the plugin reporter to see if the plugin executed. Instead, the "Go" button could be deactivated if required fields have not been completed.

### Why Should This Be In Core?

It is a core feature.

### Benefits

Users will be prevented from executing plugins with missing required data ensuring that they do not have to recommence the plugin execution process from the beginning. 

### Verification Process

Execute a graph to a GeoJSON and remove the text from the file location parameter. File > Export > to GeoJSON.
Note the execute button becomes deactivated.

### Applicable Issues
#1930

### CODE REVIEW

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/FileParameterType.java [hunk 1]
```diff
@@ -213,6 +220,8 @@ public static class FileParameterValue extends ParameterValue {
         private FileParameterKind kind;
         private ExtensionFilter filter;
         private boolean acceptAllFileFilterUsed;
+        private static final Logger LOGGER = Logger.getLogger(FileParameterValue.class.getName());
```
Comment (antares1470) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/FileParameterType.java:223]: removed unused (+ dependency)

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/FileParameterType.java [hunk 2]
```diff
@@ -343,7 +352,16 @@ public void enableAcceptAllFileFilter() {
 
         @Override
         public String validateString(final String s) {
-            return null;
+            File validationFile = new File(s);
+            //Accepted Conditions:
+            //The current file is a directory
+            //The current file has no parent
+            //the current file has a prent that is a directory
+            if (validationFile.isDirectory() || (!validationFile.isDirectory() && validationFile.getParentFile() != null && validationFile.getParentFile().exists())){
```
Comment (antares1470) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/FileParameterType.java:360]: The first bit of the clause after || is redundant as this is implied by the first condition

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/FileParameterType.java [hunk 3]
```diff
@@ -343,7 +352,16 @@ public void enableAcceptAllFileFilter() {
 
         @Override
         public String validateString(final String s) {
-            return null;
+            File validationFile = new File(s);
```
Comment (antares1470) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/FileParameterType.java:355]: final

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/FileParameterType.java [hunk 4]
```diff
@@ -343,7 +352,16 @@ public void enableAcceptAllFileFilter() {
...
+            //Accepted Conditions:
+            //The current file is a directory
+            //The current file has no parent
+            //the current file has a prent that is a directory
+            if (validationFile.isDirectory() || (!validationFile.isDirectory() && validationFile.getParentFile() != null && validationFile.getParentFile().exists())){
+                return null;
+            } else {
+                return "The specified file path doe not contain valid directories";   
+            }
```
Comment (antares1470) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/FileParameterType.java:364]: can you make this a ternary statement?

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/gui/filechooser/ShowFileChooserDialog.java [hunk 5]
```diff
@@ -67,10 +67,13 @@ public void run() {
                     final JFileChooser jfc = fileChooserBuilder.createFileChooser();
                     final FileObject currentFileObject = getCurrentFileObject();
                     if (currentFileObject != null) {
-                        jfc.setSelectedFile(new File(currentFileObject.getNameExt()));
+                        jfc.setSelectedFile(new File(currentFileObject.getName()));
+                    }
+                    int returnState = jfc.showSaveDialog(WindowManager.getDefault().getMainWindow());
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/gui/filechooser/ShowFileChooserDialog.java:72]: final

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/DataAccessPane.java [hunk 6]
```diff
@@ -318,6 +319,11 @@ public void hierarchicalUpdate() {
     public void validityChanged(boolean enabled) {
```
Comment (Delphinus8821) [inline CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/DataAccessPane.java:319]: final

File: CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/DataAccessPane.java [hunk 7]
```diff
@@ -318,6 +319,11 @@ public void hierarchicalUpdate() {
...
         // Must be overriden to implement PluginParametersPaneListener
     }
+    
+    @Override
+    public void notifyParameterValidityChange(PluginParameter<?> parameter, boolean currentlySatisfied) {
```
Comment (Delphinus8821) [inline CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/DataAccessPane.java:324]: finals

File: CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/DataSourceTitledPane.java [hunk 8]
```diff
@@ -398,4 +397,10 @@ private boolean hasNonGlobalParameters(final PluginParameters pp) {
 
         return !paramNames.isEmpty();
     }
+
+    @Override
+    public void notifyParameterValidityChange(PluginParameter<?> parameter, boolean currentlySatisfied) {
```
Comment (Delphinus8821) [inline CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/DataSourceTitledPane.java:402]: finals

File: CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/HeadingPane.java [hunk 9]
```diff
@@ -167,4 +168,9 @@ private Shape makeSquare(final Color color, final Color border) {
     protected FlowPane getBoxes() {
         return boxes;
     }
+
+    @Override
+    public void notifyParameterValidityChange(PluginParameter<?> parameter, boolean currentlySatisfied) {
```
Comment (Delphinus8821) [inline CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/HeadingPane.java:173]: finals

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/panes/HeadingPaneNGTest.java [hunk 10]
```diff
@@ -176,6 +177,11 @@ public void hierarchicalUpdate() {
             didHierarchicalUpdate = true;
         }
 
+        @Override
+        public void notifyParameterValidityChange(PluginParameter<?> parameter, boolean currentlySatisfied) {
```
Comment (Delphinus8821) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/panes/HeadingPaneNGTest.java:181]: finals

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/gui/PluginParametersSwingDialog.java [hunk 11]
```diff
@@ -223,6 +221,27 @@ private DialogDescriptor createDialogDescriptor(final boolean focused) {
 
         return dd;
     }
+    
+        
+    @Override
+    public void validityChanged(boolean valid) {
```
Comment (Delphinus8821) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/gui/PluginParametersSwingDialog.java:227]: final

**Review by Delphinus8821 — CHANGES_REQUESTED** [review]

**Review by antares1470 — APPROVED** [review]

**Review by Delphinus8821 — APPROVED** [review]

**Pull request merged**